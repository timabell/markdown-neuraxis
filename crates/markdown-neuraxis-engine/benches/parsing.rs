use criterion::{Criterion, criterion_group, criterion_main};
use pulldown_cmark::Parser;
mod common;

fn bench_pulldown_cmark_baseline(c: &mut Criterion) {
    let mut group = c.benchmark_group("parsing");
    group.sample_size(10);

    let content = common::generate_markdown_content(100);
    group.bench_function("pulldown_cmark", |b| {
        b.iter(|| {
            let parser = Parser::new(std::hint::black_box(&content));
            let events: Vec<_> = parser.collect();
            std::hint::black_box(events);
        });
    });

    group.finish();
}

criterion_group!(benches, bench_pulldown_cmark_baseline);
criterion_main!(benches);
