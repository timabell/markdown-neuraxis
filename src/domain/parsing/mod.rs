pub mod markdown;
pub mod outline;

pub use markdown::*;
pub use outline::*;
