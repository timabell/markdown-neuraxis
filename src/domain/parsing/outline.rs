use crate::domain::models::OutlineItem;

/// Build a hierarchical outline from flat list of outline items
pub fn build_hierarchy(items: Vec<OutlineItem>) -> Vec<OutlineItem> {
    let mut result = Vec::new();
    let mut stack: Vec<OutlineItem> = Vec::new();
    let mut pending_children: Vec<OutlineItem> = Vec::new();

    for item in items {
        // If this is a child item, store it for later
        if item.level > 0 {
            pending_children.push(item);
            continue;
        }

        // This is a top-level item - add any pending children to the last parent
        if let Some(mut parent) = stack.pop() {
            parent.children = pending_children;
            pending_children = Vec::new();
            result.push(parent);
        }

        stack.push(item);
    }

    // Handle the last parent with any remaining children
    if let Some(mut parent) = stack.pop() {
        parent.children = pending_children;
        result.push(parent);
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_build_flat_hierarchy() {
        let items = vec![
            OutlineItem::new("Item 1".to_string(), 0),
            OutlineItem::new("Item 2".to_string(), 0),
        ];

        let result = build_hierarchy(items);
        assert_eq!(result.len(), 2);
        assert_eq!(result[0].content, "Item 1");
        assert_eq!(result[1].content, "Item 2");
    }

    #[test]
    fn test_build_nested_hierarchy() {
        let items = vec![
            OutlineItem::new("Parent".to_string(), 0),
            OutlineItem::new("Child 1".to_string(), 1),
            OutlineItem::new("Child 2".to_string(), 1),
        ];

        let result = build_hierarchy(items);
        assert_eq!(result.len(), 1);
        assert_eq!(result[0].content, "Parent");
        assert_eq!(result[0].children.len(), 2);
        assert_eq!(result[0].children[0].content, "Child 1");
        assert_eq!(result[0].children[1].content, "Child 2");
    }
}
