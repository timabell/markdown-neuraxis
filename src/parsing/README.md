# Parsing Module

## Tree Building Challenge with pulldown-cmark

### The Problem

pulldown-cmark emits a flat stream of events for nested structures. For a nested list like:
```markdown
- Parent
  - Child
```

The event sequence is:
```
Start(List)
  Start(Item)
    Text("Parent")
    Start(List)  ← nested list comes AFTER parent text
      Start(Item)
        Text("Child")
      End(Item)
    End(List)
  End(Item)      ← parent item ends AFTER its nested list
End(List)
```

### The Challenge in Rust

The fundamental issue: when we see `Start(List)` for a nested list, we need to attach it to its parent item, but that parent item hasn't been fully constructed yet (we haven't seen `End(Item)`).

Classic solutions that don't work well:
- **Mutable reference stack**: Rust's borrow checker prevents storing `&mut` references that persist across loop iterations
- **Direct tree building**: Can't modify a parent while building its children

### Our Solution: Deferred Item Addition

Key insight: A nested list always appears between a parent item's text and its `End(Item)` event.

Strategy:
1. When we see `End(Item)`, create the item but DON'T add it yet - store as `pending_item`
2. When we see `Start(List)`:
   - If there's a `pending_item`, this list belongs to it
   - If not, it's a new top-level list
3. When we see `Start(Item)` or `End(List)`, add any `pending_item` to its parent

This works because we defer the parent-child connection until we have all the information.

### Alternative Approaches Considered

1. **Arena Allocator** (e.g., `indextree`): Good for complex trees but overkill for markdown lists
2. **Two-pass**: Parse flat with depths, then build tree - inefficient
3. **Recursive Descent**: Incompatible with pulldown-cmark's iterator API
4. **Interior Mutability** (`Rc<RefCell<>>`): Would work but adds complexity

### Implementation Notes

The current implementation uses:
- `list_stack: Vec<Vec<ListItem>>` - Stack of lists being built at each nesting level
- `pending_item: Option<ListItem>` - Item waiting to be added once we know if it has children
- `list_type_stack: Vec<bool>` - Track whether each list level is numbered or bulleted

This approach is idiomatic Rust: no unsafe code, no complex lifetimes, just careful state management.