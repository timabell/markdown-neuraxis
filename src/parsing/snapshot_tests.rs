//! Snapshot tests for markdown parsing.
//!
//! These tests use insta to verify that parsing produces the expected output
//! for various markdown inputs.

use super::parse_markdown;
use insta::assert_yaml_snapshot;
use rstest::rstest;
use std::path::PathBuf;

#[rstest]
#[case(
    r#"- First item
- Second item
- Third item"#,
    "simple_bullet_list"
)]
#[case(
    r#"- Parent item
  - Child item
  - Another child
- Second parent"#,
    "nested_bullet_list"
)]
#[case("- Single item", "single_item")]
#[case("", "empty_markdown")]
#[case(
    r#"- Item 1
  - Item 1.1
    - Item 1.1.1
      - Item 1.1.1.1
      - Item 1.1.1.2
    - Item 1.1.2
  - Item 1.2
    - Item 1.2.1
- Item 2
  - Item 2.1
- Item 3"#,
    "deep_nested_list"
)]
#[case(
    "This paragraph has [[Simple-Link]] and [[Folder/Page]] references.",
    "wiki_links_paragraph"
)]
#[case(
    r#"# Heading

Paragraph with [[Getting-Started]] link.

- List item with [[journal/2024-01-15]] reference
- Another item with [[1_Projects/Website]] link"#,
    "wiki_links_mixed"
)]
fn test_document_parsing_snapshots(#[case] markdown: &str, #[case] name: &str) {
    let doc = parse_markdown(markdown, PathBuf::from("test.md"));
    assert_yaml_snapshot!(name, doc.content);
}
