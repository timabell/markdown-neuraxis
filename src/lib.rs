pub mod app;
pub mod domain;
pub mod infrastructure;
pub mod presentation;

// Re-export commonly used types for backwards compatibility
pub use domain::models::{Document, OutlineItem};
pub use domain::parsing::PulldownMarkdownParser;

// Legacy function for backwards compatibility with existing tests
pub fn parse_markdown_outline(markdown: &str) -> Document {
    use std::path::PathBuf;
    use domain::parsing::MarkdownParser;
    let parser = PulldownMarkdownParser::new();
    parser.parse(markdown, PathBuf::from("test.md"))
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_yaml_snapshot;
    use pretty_assertions::assert_eq;
    use rstest::rstest;

    #[rstest]
    #[case("- First item\n- Second item\n- Third item", "simple_bullet_list")]
    #[case(
        "- Parent item\n  - Child item\n  - Another child\n- Second parent",
        "nested_bullet_list"
    )]
    #[case("- Single item", "single_item")]
    #[case("", "empty_markdown")]
    fn test_outline_parsing_snapshots(#[case] markdown: &str, #[case] name: &str) {
        let doc = parse_markdown_outline(markdown);
        assert_yaml_snapshot!(name, doc.outline);
    }

    #[test]
    fn test_simple_bullet_list_properties() {
        let markdown = "- First item\n- Second item\n- Third item";
        let doc = parse_markdown_outline(markdown);

        assert_eq!(doc.outline.len(), 3);
        assert_eq!(doc.outline[0].content, "First item");
        assert_eq!(doc.outline[0].level, 0);
        assert_eq!(doc.outline[1].content, "Second item");
        assert_eq!(doc.outline[2].content, "Third item");
    }

    #[test]
    fn test_nested_bullet_list_properties() {
        let markdown = "- Parent item\n  - Child item\n  - Another child\n- Second parent";
        let doc = parse_markdown_outline(markdown);

        assert_eq!(doc.outline.len(), 2);
        assert_eq!(doc.outline[0].content, "Parent item");
        assert_eq!(doc.outline[0].level, 0);
        assert_eq!(doc.outline[0].children.len(), 2);
        assert_eq!(doc.outline[0].children[0].content, "Child item");
        assert_eq!(doc.outline[0].children[0].level, 1);
        assert_eq!(doc.outline[1].content, "Second parent");
        assert_eq!(doc.outline[1].level, 0);
    }
}
