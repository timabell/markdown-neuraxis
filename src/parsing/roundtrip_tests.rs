//! Roundtrip tests for markdown parsing and generation.
//!
//! These tests ensure that content can be converted to markdown and back
//! without losing information.

use crate::models::ContentBlock;
use crate::parsing::from_markdown;

#[test]
fn test_roundtrip_conversion() {
    let original = ContentBlock::Heading {
        level: 1,
        text: "Main Title".to_string(),
    };
    let markdown = original.to_markdown();
    let converted = from_markdown(&markdown).unwrap();
    assert_eq!(original, converted);
}

#[test]
fn test_roundtrip_heading_levels() {
    for level in 1..=6 {
        let original = ContentBlock::Heading {
            level,
            text: format!("Heading Level {}", level),
        };
        let markdown = original.to_markdown();
        let converted = from_markdown(&markdown).unwrap();
        assert_eq!(original, converted);
    }
}

#[test]
fn test_roundtrip_bullet_list() {
    let original = ContentBlock::BulletList {
        items: vec![
            crate::models::ListItem::new("First item".to_string(), 0),
            crate::models::ListItem::new("Second item".to_string(), 0),
            crate::models::ListItem::new("Third item".to_string(), 0),
        ],
    };
    let markdown = original.to_markdown();
    let converted = from_markdown(&markdown).unwrap();
    assert_eq!(original, converted);
}

#[test]
fn test_roundtrip_numbered_list() {
    let original = ContentBlock::NumberedList {
        items: vec![
            crate::models::ListItem::new("First item".to_string(), 0),
            crate::models::ListItem::new("Second item".to_string(), 0),
            crate::models::ListItem::new("Third item".to_string(), 0),
        ],
    };
    let markdown = original.to_markdown();
    let converted = from_markdown(&markdown).unwrap();
    assert_eq!(original, converted);
}

#[test]
fn test_roundtrip_code_block() {
    let original = ContentBlock::CodeBlock {
        language: Some("rust".to_string()),
        code: "fn main() {\n    println!(\"Hello, world!\");\n}".to_string(),
    };
    let markdown = original.to_markdown();
    let converted = from_markdown(&markdown).unwrap();
    assert_eq!(original, converted);
}

#[test]
fn test_roundtrip_quote() {
    let original = ContentBlock::Quote("This is a quote".to_string());
    let markdown = original.to_markdown();
    let converted = from_markdown(&markdown).unwrap();
    assert_eq!(original, converted);
}

#[test]
fn test_roundtrip_rule() {
    let original = ContentBlock::Rule;
    let markdown = original.to_markdown();
    let converted = from_markdown(&markdown).unwrap();
    assert_eq!(original, converted);
}

#[test]
fn test_roundtrip_paragraph_with_wiki_links() {
    let original = ContentBlock::Paragraph {
        segments: vec![
            crate::models::TextSegment::Text("This is a ".to_string()),
            crate::models::TextSegment::WikiLink {
                target: "test-link".to_string(),
            },
            crate::models::TextSegment::Text(" paragraph.".to_string()),
        ],
    };
    let markdown = original.to_markdown();
    let converted = from_markdown(&markdown).unwrap();
    assert_eq!(original, converted);
}
