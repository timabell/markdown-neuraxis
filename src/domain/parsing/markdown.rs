use crate::domain::models::{Document, OutlineItem};
use pulldown_cmark::{Event, Parser, Tag, TagEnd};
use std::path::PathBuf;

pub trait MarkdownParser: Send + Sync {
    fn parse(&self, content: &str, path: PathBuf) -> Document;
}

#[derive(Debug, Default)]
pub struct PulldownMarkdownParser;

impl PulldownMarkdownParser {
    pub fn new() -> Self {
        Self::default()
    }
}

impl MarkdownParser for PulldownMarkdownParser {
    fn parse(&self, content: &str, path: PathBuf) -> Document {
        let parser = Parser::new(content);
        let mut items: Vec<OutlineItem> = Vec::new();
        let mut text_stack: Vec<String> = Vec::new();
        let mut list_stack: Vec<usize> = Vec::new();
        let mut in_item = false;

        for event in parser {
            match event {
                Event::Start(Tag::List(_)) => {
                    list_stack.push(0);
                }
                Event::End(TagEnd::List(_)) => {
                    list_stack.pop();
                }
                Event::Start(Tag::Item) => {
                    text_stack.push(String::new());
                    in_item = true;
                }
                Event::Text(text) if in_item => {
                    if let Some(current_text) = text_stack.last_mut() {
                        current_text.push_str(&text);
                    }
                }
                Event::End(TagEnd::Item) => {
                    in_item = false;
                    if let Some(text) = text_stack.pop() {
                        if !text.trim().is_empty() {
                            let level = list_stack.len().saturating_sub(1);
                            let item = OutlineItem::new(text.trim().to_string(), level);
                            items.push(item);
                        }
                    }
                }
                _ => {}
            }
        }

        let outline = super::outline::build_hierarchy(items);
        Document::with_outline(path, outline)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;

    #[test]
    fn test_parse_simple_markdown() {
        let parser = PulldownMarkdownParser::new();
        let content = "- First item\n- Second item";
        let doc = parser.parse(content, PathBuf::from("/test.md"));

        assert_eq!(doc.outline.len(), 2);
        assert_eq!(doc.outline[0].content, "First item");
        assert_eq!(doc.outline[1].content, "Second item");
    }

    #[test]
    fn test_parse_nested_markdown() {
        let parser = PulldownMarkdownParser::new();
        let content = "- Parent\n  - Child";
        let doc = parser.parse(content, PathBuf::from("/test.md"));

        assert_eq!(doc.outline.len(), 1);
        assert_eq!(doc.outline[0].content, "Parent");
        assert_eq!(doc.outline[0].children.len(), 1);
        assert_eq!(doc.outline[0].children[0].content, "Child");
    }
}
