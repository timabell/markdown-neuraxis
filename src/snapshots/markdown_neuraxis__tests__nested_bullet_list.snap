---
source: src/lib.rs
expression: doc.outline
---
- content: Parent item
  level: 0
  children:
    - content: Child item
      level: 1
      children: []
      metadata: {}
    - content: Another child
      level: 1
      children: []
      metadata: {}
  metadata: {}
- content: Second parent
  level: 0
  children: []
  metadata: {}
